### PR TITLE
Make `OmegaConfigLoader` accept paths containing dots outside of conf_source

### LINKED ISSUE #3004: Fix Configuration Loader to Handle Paths Containing dots
## Description

When working within Databricks workflows, it's possible to create a job that utilizes a notebook hosted inside a repository. Each time the job is executed, the code is automatically cloned to a path that depends on the commit. In my case, I have hosted a notebook that runs the Kedro project within the `.azuredevops` directory.

To locate the current project root, we can employ a regular expression (regex) pattern:

```python
pattern = r"^/Workspace/Repos/.internal/(?!./.azuredevops)."
```

With this regex, we can effectively identify the project root. For example, one possible path is:

```
/Workspace/Repos/.internal/02476ba86f_commits/20adcb96c8bec99ac2fad8b78b25158e7d968fa4/
```

However, there's an issue that arises when a period (dot) appears within the path. This problem is not limited to Databricks; it can also occur locally. It's important to note that the OmegaConfigLoader did not encounter any issues with this path until Kedro version 0.18.12.

## Context
Just running a normal kedro project. Which was working untill 0.18.12

## Steps to Reproduce
1. modify you project directory adding some `.`, for example if you a have a kedro project in `Users/yourname/kedro_project` you can create a new folder with some point `Users/yourname/.somefolder/kedro_project` 
2. [Second Step] Enable OmageConfigLoader
3. [And so on...]  Kedro run

## Expected Result
kedro should load the session

## Actual Result
The config loader is not be able to find any config file

```
`No files of YAML or JSON format found in /Workspace/Repos/.internal/02476ba86f_commits/20adcb96c8bec99ac2fad8b78b25158e7d968fa4/conf/base or /Workspace/Repos/.internal/02476ba86f_commits/20adcb96c8bec99ac2fad8b78b25158e7d968fa4/conf/databricks_dev matching the glob pattern(s): ['catalog*', 'catalog*/**', '**/catalog*']`
```

## Your Environment

* Kedro version used (`pip show kedro` or `kedro -V`): 0.18.13
* Python version used (`python -V`): 3.10.12
* Operating system and version: mac

### PR DESCRIPTION
**Pull request by ankatiyar**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

This PR is to fix #3004 

- Changes in #2977 which was a bug fix to let `OmegaConfigLoader` ignore hidden files inside the `conf/` folder (or `conf_source`) created this issue with Kedro ignoring files entirely if any part of the path contains a dot.

## Development notes

- Updated `_is_hidden` function to only return `True` when the hidden folder/file is inside the `conf_source`
- Added a test case

EDITED TO ADD:

- Turns out `pathlib.is_relative_to()` and `pathlib.relative_to()` are methods that do not work for python <3.9. And, they're already deprecated and will be removed for 3.12.
- `os.path.rel_path()` is an alternate for this functionality but it works slightly differently from `pathlib.Path.relative_to()`. So some of the relative paths contain stuff like `../../some/path` which doesn't pass the "is_hidden()" test

#### Implementation info
* I first tried to remove `resolve()` since the `path` for files that is passed to this function is already a relative path (and not do any of the resolving with `conf_path`. But this doesn't work if something like `"../user1/catalog*"` has been added to `config_patterns` (as is the case in one of the tests)
```
    def _is_hidden(self, path: str):
        """Check if path contains any hidden directory or is a hidden file"""
        path = Path(path).as_posix()
        parts = path.split(self._fs.sep)  # filesystem specific separator
        HIDDEN = "."
        # Check if any component (folder or file) starts with a dot (.)
        return any(part.startswith(HIDDEN) for part in parts)
```
* Relative path resolving with `conf_path` trips up incase the `conf_source` is a zip/tar file. So if the protocol is zip/tar, it simply checks the "relative" path passed on to this function without `resolve()`

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Review by merelcht — COMMENTED** [review]
The approach looks good to me, but I see there's some issues with the tests. I'll have another look when they're fixed 🙂 Also don't forget to add this to the release notes!

**Comment by ankatiyar** [comment]
@merelcht The CI failures are because turns out the methods I used here `path.is_relative_to()/relative_to()` are not available for lower python versions(<3.9).

File: kedro/config/omegaconf_config.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import io
 import logging
 import mimetypes
+import os
```
Comment (merelcht) [inline kedro/config/omegaconf_config.py:9]: We actually try to not use `os` if it's possible.. Is there no other way to achieve this with using `pathlib`?

**Review by merelcht — APPROVED** [review]
Thanks for fixing this and explaining the approach so well. There's definitely a lot of small edge cases to account for. 👍

**Review by SajidAlamQB — APPROVED** [review]
Thank you @ankatiyar!

**Pull request merged**